### DISCUSSION TITLE
Tidal SDK for Android Login problem

### DISCUSSION THREAD

**Discussion by ForMiSoft** [tidal-music/discussions#103] [Q&A] [answered]
I'm trying to login to Tidal in my Android application using Tidal SDK for Android. After receiving this loginUrl from initializeLogin():

`https://login.tidal.com/authorize?redirect_uri=&scope=playlist.read&client_id=<MY-CLIENT-ID>&code_challenge_method=S256&code_challenge=<CHALLENGE-CODE>&lang=en_US&appMode=android`

I opened the loginUrl in the WebView and instead of the login page it shows this and the response status is 400 (Bad Request):
![Image5](https://github.com/user-attachments/assets/5b87d6c5-0be6-4829-8df6-e699cda5d611)

Please help me to solve this problem. What am I doing wrong?

**Comment by aliaksei-taliuk** [comment]
@ForMiSoft hi! thanks for reaching out to us. I forwarded your request to the people who are responsible for Auth SDK. 
Will keep you posted.

**Reply by michpohl** [reply]
@ForMiSoft Hi!
It looks like you didn't provide a `RedirectUri` or it is an empty string. Take a look at [`Auth's `initializeLogin` function](https://github.com/tidal-music/tidal-sdk-android/blob/127085338619bab6d540bedf05f7d34a3cd1a9da/auth/src/main/kotlin/com/tidal/sdk/auth/Auth.kt#L30):
```
/**
     * Begins the login process by generating a URI to the login service.
     * @param redirectUri The URI to redirect to after successful login.
     * @param loginConfig Optional configuration for customizing the login process.
     * @return A URI to direct the user to for login.
     */
    fun initializeLogin(redirectUri: String, loginConfig: LoginConfig?): Uri {
        val uriString = loginRepository.getLoginUri(redirectUri, loginConfig)
        return Uri.parse(uriString).also {
            logger.d {
                "initializeLogin: redirectUri: $redirectUri, " +
                    "loginConfig: $loginConfig, returned uri: $it"
            }
        }
    }
```
Here you will have to submit the uri as set in the developer portal. If you don't submit the exact uri you specified there the process will fail like this.

Please report back if this helps you.

**Reply by ForMiSoft** [reply]
Thanks for your reply.
I don't know what Uri I should use. It's an Android app, not a website.

**Reply by ForMiSoft** [reply]
Hmm, I used the URL of my website, which has nothing to do with my Android app, and it worked. Now the app shows the login page. It looks like the redirectUri is unnecessary but still required. Anyway, thanks for your help—it pointed me in the right direction.

Edit:
I even did a logging of all accesses on my website, only to find out that the redirectUri is not being used in the case of the Android app at all.

**Reply by ForMiSoft** [reply]
After entering my email and password in the Tidal login dialog it shows a checkmark icon like if everything is ok, but in the app after calling auth.finalizeLogin(uri.toString()) the credentialsProvider.isUserLoggedIn() function still returns false.
![Screenshot_20241010_165912_Smart Tidal Playlist](https://github.com/user-attachments/assets/1eb2daf0-60e0-4098-9d80-7c6df0b77b04)

Any ideas what could be wrong this time?

**Reply by michpohl** [reply]
Sorry that  this is so rocky for you!
There could be a number of issues causing this. Are you interecepting the response from the login? If not, take a look at our [ComposeWebview](https://github.com/tidal-music/tidal-sdk-android/blob/main/auth/apps/demo/src/main/kotlin/com/tidal/sdk/demo/webview/ComposeWebView.kt) in the auth testing app. It uses an `ExtendedWebClient` - a `WebViewClient` that can react to receiving the correct url from the backend, to trigger calling `auth.finalizeLogin` with it. If you do not use  that url received from the backend when calling this function, it is probably the problem.
If you do, and still have issues, we might have to start looking at logs to understand what the issue is.

**Reply by ForMiSoft** [reply]
Thanks, my app is practically your auth demo app, all logic, navigation and UI is copied from that app. I will add my functionality when the authorization flow is fully functional. So far I just removed some useless (for me) stuff like the device login, updated everything to their newest versions and fixed all warnings. So yes, I'm using ComposeWebView and as I wrote in my previous post, after calling auth.finalizeLogin(uri.toString()) with the returned success uri, the credentialsProvider.isUserLoggedIn() function still returns false.

![login](https://github.com/user-attachments/assets/8613572e-9987-48f2-aece-297de1c89580)

![finalizeLogin](https://github.com/user-attachments/assets/5b5470cc-980c-4400-a478-95dfe2da0b79)

![isUserLoggedIn](https://github.com/user-attachments/assets/5614f67d-3503-4bc9-826f-f1f5b0c3f7cf)

**Reply by michpohl** [reply]
Can you log out the requests and the responses and post that here?

**Reply by michpohl** [reply]
Update: You might have uncovered a string handling issue in our system, and if this change described below helps you log in, that would prove my suspicion.

Please try the following, if your redirect url is just a domain name:
Change your redirectUri from `https://someurl.com` to `https://someurl.com/abc` or something similar like that.
After all, as you correctly observed:
> It looks like the redirectUri is unnecessary but still required.

This means for an Android app, you can basically enter whatever you want, as backend just compares with the url stored in your account, but you never need it for actual loading of anything.

**Reply by ForMiSoft** [reply]
I changed the redirectUri from `https://formisoft.cz` to `https://formisoft.cz/test` on the developer portal and in my code and it's still the same. So it wasn't the culprit. I would guess that there is a problem either with the finalizing login on the server side or with the `credentialsProvider.isUserLoggedIn()` function. Because I don't know what those functions internally do, it's hard to say which one isn't working properly.

> Can you log out the requests and the responses and post that here?

Of course, what filter should I use?

**Reply by ForMiSoft** [reply]
Wouldn't it be easier if you tried putting my project credentials in the auth demo app?

**Reply by michpohl** [reply]
I found out there actually is a bug - so you should keep using `https://formisoft.cz/test` as your redirectUri. There is an error when handling the address if there is no path after the domain name.

Make sure to completely wipe the data and do a re-sync in Android Studio just in case it before testing.

Regarding sending us credentials, I will first have to find out if I am legally allowed to do that 😄 

So for the time being, the easiest is if a) you fork the repo yourself and insert the credentials or b) turn on logging. This can be done by following this:
1. Add these dependencies to your app:
```
    implementation("org.slf4j:slf4j-api:2.0.12")
    implementation("com.github.tony19:logback-android:3.0.0")
```
2. Ad a `logback.xml` with the following contents to your `assets` folder
```
<configuration
    xmlns="https://tony19.github.io/logback-android/xml"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="https://tony19.github.io/logback-android/xml https://cdn.jsdelivr.net/gh/tony19/logback-android/logback.xsd"
>
    <appender name="logcat" class="ch.qos.logback.classic.android.LogcatAppender">
        <tagEncoder>
            <pattern>%logger{12}</pattern>
        </tagEncoder>
        <encoder>
            <pattern>[%thread] %-5level %logger{36}: %msg%n</pattern>
        </encoder>
    </appender>

    <root level="TRACE">
        <appender-ref ref="logcat" />
    </root>
</configuration>
```
3. Add this parameter when creating your `AuthConfig`:
```
 logLevel = NetworkLogLevel.BODY,

```

Then, if you filter for `tidal.com`, you should get some meaningful logs that help us understand the issue.

**Reply by ForMiSoft** [reply]
> Regarding sending us credentials, I will first have to find out if I am legally allowed to do that 😄

It's not such a big deal, I would delete that project afterwards and create a new one. And I'm pretty sure that any credentials would do the same. You can just create a new project, set is as Android one and enable its scope for playlists.read. Then you could debug it much easier than using this github proxy.

Meanwhile, I'll try the b) option.

**Reply by ForMiSoft** [reply]
I added you as a collaborator to my repository.

**Comment by ForMiSoft** [comment]
I can't send the log here, because it contains the credentials. Should I send it to your email?

**Reply by michpohl** [reply]
> I added you as a collaborator to my repository.

This was a good idea. I'll have a look in a bit.

**Reply by michpohl** [reply, marked as answer]
Here's your problem:
In `ExtendedWebClient` you changed this function:
```
private fun hasRedirectUri(uri: Uri): Boolean {
        return if (uri.toString().startsWith("")) {
            onRedirectUriReceived(uri)
            true
        } else {
            false
        }
    }
    ```
Now this one is called with every url the webview tries to load, therefore this check needs to verify if the one passed in is your `RedirectUri`. With your change this will call `onredirectUriReceived` on every address you pass in, and since it loads others before the relevant one your login process fails as the code behind `finalizeLogin` only understands the correct uri.
So if instead you check for `https://formisoft.cz/test` in here your login succeeds.

**Reply by ForMiSoft** [reply]
Thank you very much for your help. 👍

**Reply by michpohl** [reply]
Absolutely, you're welcome. Please keep in mind that there is an issue with the redirectUri needing a path (will be fixed in the next minor version, but until then, keep the `/test` or something similar in your uri) 😄

**Comment by ForMiSoft** [comment]
I closed that thread a bit prematurely. In the end it's not logged in. But
now I have another problem because the developer's portal is broken and I
can't edit my projects.

Dne út 15. 10. 2024 16:13 uživatel Michael Pohl ***@***.***>
napsal:
